### ISSUE TITLE
Whitelisting WebGL per site?

### ISSUE THREAD

**Issue by EchoDev** [ghacksuserjs/ghacks-user.js#293]
Is it possible to whitelist WebGL per site? I tried [WebAPI Manager](https://addons.mozilla.org/firefox/addon/webapi-manager/) but https://browserleaks.com/webgl is still able to detect and use WebGL when blocking WebGL Specification. I don't think I ever used WebGL in Firefox but I think it still would be a better solution if there was a proper addon to whitelist WebGL usage per site when needed instead of fully blocking WebGL.

uMatrix does not seem to support WebGL blocking, and I switched from NoScript to uMatrix recently.

And ideas?

**Comment by Thorin-Oakenpants** [comment]
I haven't looked much at WebAPI Manager ([github link](https://github.com/snyderp/web-api-manager)) because it requires all sites to have cookies enabled (for now) in order to `inject` itself (via a cookie).

I would ask @snyderp , it may be a bug

**Comment by EchoDev** [comment]
Hmm that seems to be the case, it does actually block WebGL from running in some case. The problem is that via browserleaks you can still fingerprint users. It seems like once you enable Javascript the server is able to generate the hash. Disabling WebGL would truly be the only way to stop this from happening.

**Comment by snyderp** [comment]
@EchoDev Would you mind pointing me to the tracking via WebGL?  I'm curious what end points its using for the finger printing.  Its not impossible that I missed a WebGL related method when I stubbed out all the standards (there are ~1.7k methods documented at this point in WebAPI Manager, so not unlikely one got missed :)

**Comment by EchoDev** [comment]
@snyderp according to browserleaks:

here are two kinds of WebGL fingerprints you can see:

    WebGL Report Hash

    A hash of entire WebGL Browser Report table, which contains a listing of all above WebGL capabilities and supported extensions. The hash is taken from the highest supported WebGL context dump.

    WebGL Image Hash

    A hash of the array of pixels created by drawing WebGL gradient shader. The drawing function borrowed from Fingerprintjs2 library, uniqueness percentage of generated image is currently investigated.

    The method of obtaining hash involve only WebGL API, and doesn't depend on browsers differences in toDataURL PNG export, so this fingerprint may be similar across the browsers within a single device.

**Comment by snyderp** [comment]
@EchoDev Oh geeze, thank you for pointing this out.  This seems to be a bug in the most recent version of the extension.  I'm seeing exceptions thrown by my extension on this page.  I will get this sorted out right away and post back here once I know whats up.  Thanks for the nod!

**Comment by snyderp** [comment]
@EchoDev 

More info on this: the page uses the WebGL2.0 standard, not just WebGL.  I should add that in!  I'll get it done tonight.  Thanks again for this!

**Comment by earthlng** [comment]
@snyderp I noticed another problem with your extension. If you go to https://browserleaks.com/canvas and make sure nothing is blocked, you can see that Canvas toDataURL is True and there's some stuff under Your Fingerprint. Now use Lite Settings for that site and enable Log blocked functionality, open the console and refresh the page: Canvas toDataURL is now false and Your Fingerprint is n/a but the console doesn't show anything about what's blocked. Why is that?

**Comment by earthlng** [comment]
Oh and is it maybe possible to add something that allows us to use only parts of the groups that you made available, like a Custom group or something like that? for example I would like to block only 'Element.prototype.getAttribute' from 'DOM Level 1' to stop the [Adblock Detection](https://browserleaks.com/proxy) shenanigans by default for all pages. Blocking the full DOM Level 1 will probably cause too much breakage but most sites should work fine without .getAttribute.

**Comment by snyderp** [comment]
@earthlng Ok, the new version I just pushed should (mostly) address both the issues you mentioned.  Its version 0.9.13.  Make sure you blocked the new "WebGL 2.0" standard.

You'll see more of the WebGL fingerprinting information blocked.  However, there are still some identifiers leaked.  This is because the extension only blocks methods in the web standards (I can say more about why it doesn't block properties, but it might not be of interest, and in some cases isn't possible from the extension vantage point)

Some of the WebGL information though is set on properties of the `WebGL2RenderingContext` object.  These are the ones that you'll still see after 0.9.13 on this page.  I'm not sure there is much that can be done to block these with an extension, that wouldn't cause lots of other breaking.  To do so accurately, you'd need to make C++ level changes.

Hope that helps!

**Comment by snyderp** [comment]
@earthlng Thanks for raising this issue. The behavior your describing though is intended.  The reason you don't see the block mentioned in the logs though is because the fingerprinting attempt is being done on the `contentWindow` of an injected iframe (to work around how most extensions try to block page access to this stuff).  WebAPI Manager prevents pages from accessing the DOMs of inserted frames, to prevent these kinds of attacks / circumvention attempts.  However, as a result, it means that the extension doesn't know what the page is trying to do in the inserted iframe, since it blocks all access to it.

So, TL;DR; the extension will still block attempts to access disabled standards in inserted frames, but looses the ability to log about them

**Comment by EchoDev** [comment]
@snyderp it is too bad the WebGL report hash can't be faked. The WebGL image hash was easily faked using something like CanvasBlocker but the report hash stays the same. Better than nothing I guess =)

**Comment by snyderp** [comment]
@EchoDev I think its unfortunately a specific use vs general purpose tradeoff.  The main goal of my extension is to reduce the attack surface pages have (for privacy and security purposes), so goal is very general (toggle all methods with standard level granularity, while minimizing how much unrelated code breaks), while a tool like `CanvasBlocker` (I'm guessing, I haven't used it) might go through and modify the environment beyond that to address a very specific type of attack.

For what its worth though, most of the highly identifying information is behind the methods e timing based fingerprinting attacks require methods, etc.).  The static property stuff will be general to the level of the browser / OS / graphics card.  Eitherway, hope its at least somewhat helpful :)

**Comment by snyderp** [comment]
@earthlng I thought about that, but couldn't come up with a usable interface to give users per-feature blocking granularity.  I could imagine some "experts text field" that allows you to drop in specific keypaths to block, that could be useful.  I'll open an issue now just to keep track of the idea.  

I'll be at least a while before I get to it though.  In the meantime, PR very welcome ;)

**Comment by earthlng** [comment]
Thanks for the explanation. But it makes troubleshooting a lot harder if there's no indication that WAM blocked something other than the selected standards for a page.

**Comment by Thorin-Oakenpants** [comment]
@earthlng - see https://github.com/snyderp/web-api-manager/issues/42 for tracking your request

@EchoDev - I think this is the fix for this issue re WebGL - https://github.com/snyderp/web-api-manager/issues/41

Closing this
